spec/index.bs

```diff
@@ -3787,43 +3788,45 @@ GPUQueue includes GPUObjectBase;
 
 <dl dfn-type="method" dfn-for="GPUQueue">
 <div algorithm>
-    : <dfn>writeBuffer(|sourceData|, |sourceOffset|, |destination|, |destinationOffset|, |size|)</dfn>
+    : <dfn>writeBuffer(|buffer|, |bufferOffset|, |size|, |data|, |dataOffset|)</dfn>
     ::
-        Takes the |sourceData| contents of size |size|, starting from the byte offset |sourceOffset|,
-        and schedules a write operation of these contents to the |destination| buffer on the
-        [=Queue timeline=] starting at |destinationOffset|.
-        Any subsequent modifications to |sourceData| do not affect what is written
+        Takes the |data| contents of size |size|, starting from the byte offset |dataOffset|,
+        and schedules a write operation of these contents to the |buffer| buffer on the
+        [=Queue timeline=] starting at |bufferOffset|.
+        Any subsequent modifications to |data| do not affect what is written
         at the time that the scheduled operation runs.
 
-        The operation throws {{OperationError}} if any of the following is true:
-        - |destination| buffer isn't in the `"unmapped"` [=buffer state=].
-        - |destinationOffset| is not a multiple of 4.
-        - |size| is not a positive multiple of 4.
-        - |sourceOffset| + |size| exceeds |sourceData|.byteLength.
+        The operation throws {{OperationError}} if |dataOffset| + |size| exceeds |data|.byteLength.
 
         The operation does nothing and produces an error if any of the following is true:
-        - |destination|.{{GPUBuffer/[[usage]]}} doesn't include {{GPUBufferUsage/COPY_DST}} flag.
-        - |destination| buffer is destroyed
-        - |destinationOffset| + |size| exceeds |destination|.{{GPUBuffer/[[size]]}}.
+        - |bufferOffset| is not a multiple of 4.
```

Comment:
What's the benefit of deferring this validation to the GPU/server side?
My general understanding was that if something is directly check-able (outside of the current execution context), we should do it on the client side.

Reply:
You still need to validate a second time the GPU process, and defaulting to validate things on the device timeline will avoid mistakes or issues when we extend the API: the same reasoning would lead to validating the 256 alignment of UBOs on the content timeline, but it will become a limit.

Reply:
Hmm. That's true. But also - if we validate it on the client side correctly, and it fails the validation on the server side, that means we are being attacked. It's no longer a case where we care about carefully issuing the exception and pretending that we are good. We can just panic, or do something entirely different.

I think it's beneficial to report errors earlier, if we can.

Also, as for the extensions later, we can always *remove* the client side validation - it would still be compatible with the old code (e.g. for the 256 alignment limit).

Reply:
> I think it's beneficial to report errors earlier, if we can.

Can you elaborate? I think it might help for development but I don't see any advantage for production pages.

> Also, as for the extensions later, we can always remove the client side validation - it would still be compatible with the old code (e.g. for the 256 alignment limit).

I guess that would be mostly non-breaking. I still worry that it's easier to not make mistakes (and the spec is easier to understand / implement) if most things are handled on the device timeline, and only the necessary things are in the content timeline.

Reply:
> Can you elaborate? I think it might help for development but I don't see any advantage for production pages.

I don't think there is any harm to production pages either if errors show up earlier.

As a user, when an application hits a bug somewhere, I'd find it much more useful to see "Exception: your buffer size is not aligned to 4" instantly than "Oops, something went wrong, please write to customer support" a few frames later. In the former case, I may at least have some ideas on what buffer is there, maybe it's something I passed in (by uploading a file), so I can find ways to work around it, or even just report properly to customer support.

Reply:
Ok, this is the most controversial bit, so I'll remove it and start a second PR once this one is landed so as to not block things.